### DISCUSSION TITLE
private Dataportal_xxxx legacy methods not found if they're in a base class

### DISCUSSION THREAD

**Discussion by amselem** [MarimerLLC/csla#2309] [Questions]
Hi

I'm testing migration of a Csla 4.8 application to 5.4.2.

When the dataportal is looking for the fallback methods (DataPortal_XXX), it doesn't look in the base classes, only in the target type. This is a breaking change from 4.8 as it can't find something like a private void DataPortal_XXX method in the base class.

Old code: MethodCaller.cs  , method FindMethod
(there's a loop traversing the hierarchy)

 ```
      do
      {
        // find for a strongly typed match
        info = objectType.GetMethod(method, oneLevelFlags, null, types, null);
        if (info != null)
        {
          break; // match found
        }

        objectType = objectType.BaseType();

      } while (objectType != null);
```

New code: ServiceProviderMethodCaller.cs  , FindDataPortalMethod
(There's no loop, it only takes targetType methods as candidates)
```

  // if no attribute-based methods found, look for legacy methods
        if (!candidates.Any())
        {
          var attributeName = typeOfOperation.Name.Substring(0, typeOfOperation.Name.IndexOf("Attribute"));
          var methodName = attributeName.Contains("Child") ?
              "Child_" + attributeName.Substring(0, attributeName.IndexOf("Child")) :
              "DataPortal_" + attributeName;
          candidates = targetType.GetMethods(_bindingAttr).Where(
            m => m.Name == methodName);
        }
```

Suggested correction: call the old MethodCaller code
```

  // if no attribute-based methods found, look for legacy methods
        if (!candidates.Any())
        {
          var attributeName = typeOfOperation.Name.Substring(0, typeOfOperation.Name.IndexOf("Attribute"));
          var methodName = attributeName.Contains("Child") ?
              "Child_" + attributeName.Substring(0, attributeName.IndexOf("Child")) :
              "DataPortal_" + attributeName;

          var method = MethodCaller.GetMethod(targetType, methodName, criteria);
          if(method != null) 
            candidates = new System.Reflection.MethodInfo[] { method };
```

**Comment by rockfordlhotka** [comment]
This appears to be a regression bug. We're clearly missing a test that should have caught this.

Would you be so kind as to file this as an issue in the Issues tab?

**Reply by rockfordlhotka** [reply]
Perhaps related to this one? https://github.com/MarimerLLC/csla/issues/2287

**Reply by amselem** [reply]
Yes, it looks like the same issue!   I won't file this to not duplicate the issue then.
Thank you Rocky
